File: 8.0.1/syncml/test/com/funambol/syncml/client/FileSyncSourceTest.java
/*
 * Funambol is a mobile platform developed by Funambol, Inc. 
 * Copyright (C) 2008 Funambol, Inc.
 * 
 * This program is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License version 3 as published by
 * the Free Software Foundation with the addition of the following permission 
 * added to Section 15 as permitted in Section 7(a): FOR ANY PART OF THE COVERED
 * WORK IN WHICH THE COPYRIGHT IS OWNED BY FUNAMBOL, FUNAMBOL DISCLAIMS THE 
 * WARRANTY OF NON INFRINGEMENT  OF THIRD PARTY RIGHTS.
 * 
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 * details.
 * 
 * You should have received a copy of the GNU Affero General Public License 
 * along with this program; if not, see http://www.gnu.org/licenses or write to
 * the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA 02110-1301 USA.
 * 
 * You can contact Funambol, Inc. headquarters at 643 Bair Island Road, Suite 
 * 305, Redwood City, CA 94063, USA, or at email address info@funambol.com.
 * 
 * The interactive user interfaces in modified source and object code versions
 * of this program must display Appropriate Legal Notices, as required under
 * Section 5 of the GNU Affero General Public License version 3.
 * 
 * In accordance with Section 7(b) of the GNU Affero General Public License
 * version 3, these Appropriate Legal Notices must retain the display of the
 * "Powered by Funambol" logo. If the display of the logo is not reasonably 
 * feasible for technical reasons, the Appropriate Legal Notices must display
 * the words "Powered by Funambol". 
 */

package com.funambol.syncml.client;

import java.io.IOException;
import java.io.OutputStream;
import java.io.InputStream;
import java.util.Vector;
import java.util.Enumeration;
import javax.microedition.io.Connector;
import javax.microedition.io.file.FileConnection;

import com.funambol.storage.StringKeyValueStore;
import com.funambol.storage.StringKeyValueFileStore;
import com.funambol.syncml.spds.SourceConfig;
import com.funambol.syncml.spds.SyncItem;
import com.funambol.syncml.spds.SourceConfig;
import com.funambol.syncml.spds.SyncException;
import com.funambol.syncml.protocol.SyncML;
import com.funambol.util.ConsoleAppender;
import com.funambol.util.Log;

import com.funambol.util.FunBasicTest;
import j2meunit.framework.*;

public class FileSyncSourceTest extends FunBasicTest {
    private StringKeyValueStore store;
    private FileSyncSource      source;
    private TestTracker         tracker;
    private String              directory;
    private SourceConfig        config;

    private class TestTracker implements ChangesTracker {

        private Vector newItems = new Vector();
        private Vector delItems = new Vector();
        private Vector updItems = new Vector();
        private Vector allItems = new Vector();

        public TestTracker() {
        }

        public void setSyncSource(TrackableSyncSource ss) {
        }

        public void begin() throws TrackerException {
        }

        public void end() throws TrackerException {
        }

        public Enumeration getNewItems() throws TrackerException {
            return newItems.elements();
        }

        public int getNewItemsCount() throws TrackerException {
            return newItems.size();
        }

        public Enumeration getUpdatedItems() throws TrackerException {
            return updItems.elements();
        }

        public int getUpdatedItemsCount() throws TrackerException {
            return updItems.size();
        }

        public Enumeration getDeletedItems() throws TrackerException {
            return delItems.elements();
        }

        public int getDeletedItemsCount() throws TrackerException {
            return delItems.size();
        }

        public void setItemStatus(String key, int status) throws TrackerException  {
        }

        public void reset() throws TrackerException {
        }

        public boolean removeItem(SyncItem item) throws TrackerException {
            return false;
        }

        public void addNewItem(SyncItem item) {
            newItems.addElement(item);
        }

        public void addUpdItem(SyncItem item) {
            updItems.addElement(item);
        }

        public void addDelItem(SyncItem item) {
            delItems.addElement(item);
        }

        public void addItem(SyncItem item) {
            allItems.addElement(item);
        }
    }

    public FileSyncSourceTest() throws Exception {
        super(3, "FileSyncSourceTest");

        Log.initLog(new ConsoleAppender());
        Log.setLogLevel(Log.TRACE);

        directory = "file:///root1";
    }

    public void setUp() {
        directory = "file:///root1";
    }

    private void prepareTestDirectory(String baseDir, String testDir) throws IOException {

        FileConnection fc = (FileConnection) Connector.open(baseDir + '/' + testDir,
                                                            Connector.READ_WRITE);
        if (fc.exists()) {
            Enumeration files = fc.list();
            while(files.hasMoreElements()) {
                String file = (String)files.nextElement();
                String fullFile = baseDir + '/' + testDir + '/' + file;
                FileConnection fc1 = (FileConnection) Connector.open(fullFile,
                                                                     Connector.READ_WRITE);
                try {
                    fc1.delete();
                } catch (Exception e) {
                    Log.error("Exception while deleting: " + fullFile + " -- " + e.toString());
                }
                fc1.close();
            }
        } else {
            fc.mkdir();
        }
        fc.close();

        directory = baseDir + '/' + testDir + '/';
        tracker = new TestTracker();
        config = new SourceConfig(SourceConfig.BRIEFCASE,"application/*", "briefcase");
        Log.debug("directory = " + directory);
        source = new FileSyncSource(config, tracker, directory);
    }

    public void test(int testNumber) throws Throwable {
        switch (testNumber) {
            case 0:
                testSlowSyncSimple();
                break;
            case 1:
                testFastSyncSimple();
                break;
            case 2:
                testSyncSendLargeObject();
                break;
        }
    }

    private void testSlowSyncSimple() throws Throwable {

        prepareTestDirectory(directory, "test0");
        // Populate the directory with one file
        String content = "This is a small file";
        createFile("test-0.txt", content);
        tracker.addItem(new SyncItem("test-0.txt"));

        source.beginSync(SyncML.ALERT_CODE_SLOW);
        SyncItem item = source.getNextItem();
        assertTrue(item != null);
        assertTrue(source.getNextItem() == null);
        byte itemContent[] = item.getContent();
        assertTrue(itemContent != null);
        assertTrue(content.equals(new String(itemContent)));
        source.endSync();
    }


    private void testFastSyncSimple() throws Throwable {
    }


    private void testSyncSendLargeObject() throws Throwable {
        // Populate the directory with one file
        prepareTestDirectory(directory, "test1");
        byte content[] = new byte[26*1024];
        String fileContent = new String(content);
        createFile("test-0.txt", fileContent);
        tracker.addItem(new SyncItem("test-0.txt"));

        source.beginSync(SyncML.ALERT_CODE_SLOW);
        SyncItem chunk1 = source.getNextItem();
        assertTrue(chunk1 != null);
        assertTrue(chunk1.hasMoreData());
        SyncItem chunk2 = source.getNextItem();
        assertTrue(chunk2 != null);
        assertTrue(!chunk2.hasMoreData());
        assertTrue(source.getNextItem() == null);
        assertTrue(fileContent.equals(new String(chunk1.getContent()) + new String(chunk2.getContent())));
        source.endSync();
    }

    private void createFile(String fileName, String content) throws Throwable {
        FileConnection fc = (FileConnection) Connector.open(directory + "/" + fileName,
                                                            Connector.READ_WRITE);
        OutputStream os;
        if (!fc.exists()) {
            fc.create();
            os = fc.openOutputStream();
        } else {
            // Truncate and open
            fc.truncate(0);
            os = fc.openOutputStream();
        }
        os.write(content.getBytes());
        os.close();
        fc.close();
    }
}

File: v85ga/test/common/client/FileSyncSourceTest.cpp
/*
 * Funambol is a mobile platform developed by Funambol, Inc. 
 * Copyright (C) 2008 Funambol, Inc.
 * 
 * This program is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License version 3 as published by
 * the Free Software Foundation with the addition of the following permission 
 * added to Section 15 as permitted in Section 7(a): FOR ANY PART OF THE COVERED
 * WORK IN WHICH THE COPYRIGHT IS OWNED BY FUNAMBOL, FUNAMBOL DISCLAIMS THE 
 * WARRANTY OF NON INFRINGEMENT  OF THIRD PARTY RIGHTS.
 * 
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 * details.
 * 
 * You should have received a copy of the GNU Affero General Public License 
 * along with this program; if not, see http://www.gnu.org/licenses or write to
 * the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA 02110-1301 USA.
 * 
 * You can contact Funambol, Inc. headquarters at 643 Bair Island Road, Suite 
 * 305, Redwood City, CA 94063, USA, or at email address info@funambol.com.
 * 
 * The interactive user interfaces in modified source and object code versions
 * of this program must display Appropriate Legal Notices, as required under
 * Section 5 of the GNU Affero General Public License version 3.
 * 
 * In accordance with Section 7(b) of the GNU Affero General Public License
 * version 3, these Appropriate Legal Notices must retain the display of the
 * "Powered by Funambol" logo. If the display of the logo is not reasonably 
 * feasible for technical reasons, the Appropriate Legal Notices must display
 * the words "Powered by Funambol".
 */

# include <cppunit/extensions/TestFactoryRegistry.h>
# include <cppunit/extensions/HelperMacros.h>

#include "base/fscapi.h"
#include "client/FileSyncSource.h"
#include "testUtils.h"

USE_NAMESPACE

#define TEST_INPUT_DIR      SAMPLE_FILES_DIR      // here we have input files
#define TEST_OUTPUT_DIR     "FileSyncSourceTest"

#define TEST_FILE_NAME1     "pic.jpg"
#define TEST_FILE_NAME2     "pic2.jpg"
#define TEST_FILE_NAME3     "pic3.jpg" 


class FileSyncSourceTest : public CppUnit::TestFixture {

    CPPUNIT_TEST_SUITE(FileSyncSourceTest);
    CPPUNIT_TEST(testAdd);
    CPPUNIT_TEST(testAddSameName);
    CPPUNIT_TEST(testAddSameNameSecondFile);
    CPPUNIT_TEST(testAddSameNameSecondFile2);
    CPPUNIT_TEST(testAddSameNameThirdFile);
    CPPUNIT_TEST(testAddNoName);
    CPPUNIT_TEST(testAddRawFile);
    CPPUNIT_TEST(cleanup);
    CPPUNIT_TEST_SUITE_END();

public:

    void setUp() {

        // Create the output dir
        outputDir = getTestDirFullPath(TEST_OUTPUT_DIR);
        createFolder(outputDir.c_str());

        // Create the FileSyncSource
        SyncSourceConfig fssc;
        fss = new FileSyncSource(TEXT("testFss"), &fssc, outputDir);

        ssr = new SyncSourceReport("testFss");
        fss->setReport(ssr);
    }

    void tearDown() {
        delete fss;
        delete ssr;
    }

private:

    StringBuffer outputDir;
    FileSyncSource* fss;
    SyncSourceReport* ssr;


    /// Util method: creates a OMA file data object
    FileData createFileData(const char* inputFileName) {

        // Read the input file
        char* content = NULL;
        size_t len;
        StringBuffer inFile = getTestFileFullPath(TEST_INPUT_DIR, inputFileName);
        bool fileLoaded = readFile(inFile.c_str(), &content, &len, true);
        CPPUNIT_ASSERT_MESSAGE("Failed to load test file", fileLoaded);

        FileData file;
        file.setBody(content, (int)len);
        file.setSize((int)len);
        WString wname;
        wname = inputFileName;
        file.setName(wname.c_str());

        delete [] content;
        return file;
    }


    void insertItem(const char* testFileName, const int expectedResult, 
                    const char* expectedOutName, const char* forceFileName = NULL) {

        FileData file = createFileData(testFileName);
        if (forceFileName) {
            WString wname;
            StringBuffer tmp(forceFileName);
            wname = tmp;
            file.setName(wname.c_str());
        }

        // Create a SyncItem containing the OMA file data
        SyncItem item(TEXT("test1"));
        char* data = file.format();
        item.setData(data, (long)strlen(data));
        delete [] data;

        int ret = fss->insertItem(item);

        // TESTS
        CPPUNIT_ASSERT_MESSAGE("wrong status returned", ret == expectedResult);
        if (fss->isErrorCode(ret)) {
            return;
        }
        StringBuffer outKey;
        outKey.convert(item.getKey());
        CPPUNIT_ASSERT_MESSAGE("wrong item's key returned", outKey == expectedOutName);

        StringBuffer outFile(outputDir);
        outFile.append(expectedOutName);
        CPPUNIT_ASSERT_MESSAGE("output file not found", fileExists(outFile.c_str()) );

        char* outContent = NULL;
        size_t outLen;
        readFile(outFile.c_str(), &outContent, &outLen, true);
        CPPUNIT_ASSERT( outLen == file.getSize() );
        delete [] outContent;
    }


    void insertRawItem(const char* testFileName, const int expectedResult, 
                       const WCHAR* inputItemKey, const WCHAR* expectedOutName) {


        // Read the input file
        char* content = NULL;
        size_t len;
        StringBuffer inFile = getTestFileFullPath(TEST_INPUT_DIR, testFileName);
        bool fileLoaded = readFile(inFile.c_str(), &content, &len, true);
        CPPUNIT_ASSERT_MESSAGE("Failed to load test file", fileLoaded);
        
        // Create a SyncItem containing the raw file data
        SyncItem item(inputItemKey);
        item.setData(content, (long)len);
        delete [] content;

        int ret = fss->insertItem(item);

        // TESTS
        CPPUNIT_ASSERT_MESSAGE("wrong status returned", ret == expectedResult);
        if (fss->isErrorCode(ret)) {
            return;
        }
        WString outKey(item.getKey());
        CPPUNIT_ASSERT_MESSAGE("wrong item's key returned", outKey == expectedOutName);

        StringBuffer outFile(outputDir);
        StringBuffer outName;
        outName.convert(expectedOutName);
        outFile.append(outName);
        CPPUNIT_ASSERT_MESSAGE("output file not found", fileExists(outFile.c_str()) );

        char* outContent = NULL;
        size_t outLen;
        readFile(outFile.c_str(), &outContent, &outLen, true);
        CPPUNIT_ASSERT( outLen == len );
        delete [] outContent;
    }


    //
    ///////////////////////////////////////// TESTS /////////////////////////////////////////
    //

    /**
     * 1. Adds a file TEST_FILE_NAME1, expected ok (200).
     */
    void testAdd() {
        insertItem(TEST_FILE_NAME1, STC_OK, TEST_FILE_NAME1);
    }

    /**
     * 2. Adds the same file (name & content), expected code 418.
     */
    void testAddSameName() {
        insertItem(TEST_FILE_NAME1, STC_ALREADY_EXISTS, TEST_FILE_NAME1);
    }

    /**
     * 3. Adds a second file with same name (different content), expected ok (200) 
     * and the file name with "_01" suffix.
     */
    void testAddSameNameSecondFile() {
        StringBuffer expectedOutName(TEST_FILE_NAME1);
        expectedOutName.replace(".", "_01.");
        insertItem(TEST_FILE_NAME2, STC_OK, expectedOutName, TEST_FILE_NAME1);
    }

    /**
     * 4. Adds again the second file, expected code 418 and the file name with "_01" suffix.
     */
    void testAddSameNameSecondFile2() {
        StringBuffer expectedOutName(TEST_FILE_NAME1);
        expectedOutName.replace(".", "_01.");
        insertItem(TEST_FILE_NAME2, STC_ALREADY_EXISTS, expectedOutName, TEST_FILE_NAME1);
    }

    /**
     * 5. Adds a third file with same name (different content), expected ok (200) 
     * and the file name with "_02" suffix.
     */
    void testAddSameNameThirdFile() {
        StringBuffer expectedOutName(TEST_FILE_NAME1);
        expectedOutName.replace(".", "_02.");
        insertItem(TEST_FILE_NAME3, STC_OK, expectedOutName, TEST_FILE_NAME1);
    }

    /**
     * 6. Adds a file with no name, expected error (code 500).
     */
    void testAddNoName() {
        insertItem(TEST_FILE_NAME1, STC_COMMAND_FAILED, NULL, "");
    }

    /**
     * 7. Adds a raw file, expected ok (200).
     * Note: it's a raw file so the file name is the item's key from server.
     */
    void testAddRawFile() {
        WString itemKey(TEXT("key-from-server"));
        insertRawItem(TEST_FILE_NAME1, STC_OK, itemKey.c_str(), itemKey.c_str());
    }

    /// Cleans up the destination folder, removing the 4 files created.
    /// This should be launched at the end of all tests for FileSyncSourceTest.
    void cleanup() {
        removeFileInDir(outputDir.c_str());
    }
};

CPPUNIT_TEST_SUITE_REGISTRATION( FileSyncSourceTest );
